src/nnet3/nnet-discriminative-example.h

```diff
@@ -0,0 +1,410 @@
+                    // frames).  We may still need to include those frames in
+                    // the computation in order to get correct posteriors for
+                    // other parts of the lattice.
+
+  bool split; // if false, we won't split at all.
+
+  bool excise; // if false, we will skip the "excise" step.
+  
+  SplitDiscriminativeExampleConfig():
+      max_length(1024), criterion("smbr"), collapse_transition_ids(true),
+      determinize(true), minimize(true), test(false), drop_frames(false),
+      split(true), excise(true) { }
+
+  void Register(OptionsItf *opts) {
+
+    opts->Register("max-length", &max_length, "Maximum length allowed for any "
+                   "segment (i.e. max #frames for any example");
+    opts->Register("criterion", &criterion, "Criterion, 'mmi'|'mpfe'|'smbr'. "
+                   "Determines which frames may be dropped from lattices.");
+    opts->Register("collapse-transition-ids", &collapse_transition_ids,
+                   "This option included for debugging purposes");
+    opts->Register("determinize", &determinize, "If true, we determinize "
+                   "lattices (as Lattice) before splitting and possibly minimize");
+    opts->Register("minimize", &minimize, "If true, we push and "
+                   "minimize lattices (as Lattice) before splitting");
+    opts->Register("test", &test, "If true, activate self-testing code.");
+    // See "Sequence-discriminative training of deep neural networks", Vesely et al,
+    // ICASSP 2013 for explanation of frame dropping.
+    opts->Register("drop-frames", &drop_frames, "For MMI, if true we drop frames "
+                   "with no overlap of num and den frames");
+    opts->Register("split", &split, "Set to false to disable lattice-splitting.");
+    opts->Register("excise", &excise, "Set to false to disable excising un-needed "
+                   "frames (option included for debug purposes)");
+  }
+};
+
+/// This struct exists only for diagnostic purposes.  Note: the stats assume
+/// that you call SplitDiscriminative and ExciseDiscriminativeExample in the
+/// same program, and the info printed out will be wrong if this is not the
+/// case... this isn't ideal but it was more convenient.
+struct SplitExampleStats {
+  int32 num_lattices;
+  int32 longest_lattice;
+  int32 num_segments;
+  int32 num_kept_segments;
+  int64 num_frames_orig;
+  int64 num_frames_must_keep;
+  int64 num_frames_kept_after_split;
+  int32 longest_segment_after_split;
+  int64 num_frames_kept_after_excise;
+  int32 longest_segment_after_excise;
+  
+  SplitExampleStats() { memset(this, 0, sizeof(*this)); }
+  void Print();
+};
+
+/** Split a "discriminative example" into multiple pieces,
+    splitting where the lattice has "pinch points".
+    Uses "name" as the supervision object that is used to 
+    determine the "pinch points".
+ */
+void SplitDiscriminativeExample(
+    const std::string &name,
```

Comment:
Actually, I know I told you we should support multiple supervision objects,
but on second thoughts, I think it makes sense to allow just one (and of
course store its name).  In a multilingual setup, an utterance corresponds
to just one language.

In addition, after thinking about this a bit, I think we're going to have
to make some substantial changes from the 'nnet2' way of doing the
discriminative training.  In nnet2 we split the lattices on pinch points,
and I think there was some way of padding them and stitching examples
together.  But if we have recurrent architectures that see infinite
context, stitching examples together won't fly, and even padding at the
ends isn't quite right.  Also, there is a big cost to using variable-length
egs, because of how the compilation works.  So we will need to rely more on
fixed-length egs extracted from the lattice without regard to where the
pinch points lie.  This is what I do in the 'chain' models.  I use
fixed-length egs (1.5 seconds by default), and discard training utterances
shorter than this.  (we can append training data at the data-dir level if
we're concerned about losing too many short utterances; @tomkocse already
wrote a script for this).

So we have extract fixed-length egs from the lattices.  The edge effects
can be handled by using the 'forward' and 'backward' scores of the cut
points as the initial and final-probs.  [you can of course renormalize
somehow so the best cost is zero.]   Initial-probs can be simulated using
arc probabilities.  In order to know which frames the acoustic scores
correspond to, the decoder will have to dump in the non-compact lattice
format (--determinize=false), and because this takes up a lot of disk, we
can eventually consider integrating the decoding with the initial phase of
egs-dumping.  But for now probably best to just dump the lattices without
determinization.

The initial splitting up of lattices can decide on random fixed-length
pieces of lattice- use the 'SplitIntoRanges' function from the 'chain'
branch.  The lattice splitting-up code will be similar to class
SupervisionSplitter in the 'chain' branch, except with more attention to
the initial and final costs. (To do this, in addition to computing the
lattice state times, you'll want to compute the lattice alpha and beta
scores).

Obviously this is a slightly bigger project than we thought, now.  If you
don't have time, feel free to reconsider.

Dan


On Sat, Dec 19, 2015 at 10:24 AM, Vimal Manohar <notifications@github.com>
wrote:

> In src/nnet3/nnet-discriminative-example.h
> <https://github.com/danpovey/kaldi/pull/3#discussion_r48095338>:
>
> > +  int64 num_frames_kept_after_split;
> > +  int32 longest_segment_after_split;
> > +  int64 num_frames_kept_after_excise;
> > +  int32 longest_segment_after_excise;
> > +
> > +  SplitExampleStats() { memset(this, 0, sizeof(*this)); }
> > +  void Print();
> > +};
> > +
> > +/** Split a "discriminative example" into multiple pieces,
> > +    splitting where the lattice has "pinch points".
> > +    Uses "name" as the supervision object that is used to
> > +    determine the "pinch points".
> > + */
> > +void SplitDiscriminativeExample(
> > +    const std::string &name,
>
> If there are multiple supervisions, then only the "name" object will be
> considered to identify the pinch points.
> This is a difference from nnet2. The Excise function also has the same
> issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/danpovey/kaldi/pull/3/files#r48095338>.
>